binary_search_tree/tree.py

```diff
@@ -1,58 +1,211 @@
+            elif current.key < key:
+                 current = current.right
+            else:
+                 current = current.left
 
-    # Time Complexity: 
-    # Space Complexity: 
+#===========INORDER==============================================================
+
+    # Time Complexity:O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        result = []
+        return self.inorder_helper(self.root, result)
+    
+    def inorder_helper(self, current, result):
+        
+        if current == None:
+            return result
+        else:
+            if current.left != None:
+            #recur on the left child
+               self.inorder_helper(current.left,result)
 
-    # Time Complexity: 
-    # Space Complexity:     
+            #print the data of the node
+            result.append({"key": current.key, "value" : current.value})
+            
+            if current.right != None:
+            #recur on the right child
+                self.inorder_helper(current.right, result)
+        
+        return result
+
+
+#==============PREORDER=============================================================
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def preorder(self):
-        pass
+        result = []
+        return self.preorder_helper(self.root, result)
+    
+    def preorder_helper(self, current, result):
+        if current == None:
+            return result
+        else:
+            #print the data of the node
+            result.append({"key": current.key, "value" : current.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+            if current.left != None:
+            #recur  on the left child
+                self.preorder_helper(current.left, result)
+
+            if current.right != None:
+            #recur on the right child
+                self.preorder_helper(current.right, result)
+        
+        return result
+    
+#====================POSTORDER======================================================
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def postorder(self):
-        pass
+        result = []
+        return self.postorder_helper(self.root, result)
+    
+    def postorder_helper(self, current, result):
+        if current == None:
+            return result
+        else:
+            #recur on the left child
+            if current.left != None:
+                self.postorder_helper(current.left, result)
+            #recur on the right child
+            if self.root.right != None:
+                self.postorder_helper(current.right, result)
+            #print the data of the node
+            result.append({"key": current.key, "value" : current.value})
+        return result
 
-    # Time Complexity: 
-    # Space Complexity:     
+
+#==============HEIGHT============================================================
+    # Time Complexity: O(n)
+    # Space Complexity: O(1) 
     def height(self):
```

Comment:
👍 However since the height helper is recursive, `height` has a space complexity of O(log n)